Testing workshop TODOs (#27)
Android-Academy-Global/fundamentals-2020-homework: workshop/workshop3.md

# Workshop 3

- Now we can refactor and simplify code from Workshop 1-2

## TODO 3.1

- Open `MovieFactory.kt` file and uncomment next code:

```kotlin
fun createMovie(
    id: Int = 0,
    title: String = "",
    reviewCount: Int = 0,
    isLiked: Boolean = false,
    rating: Int = 0,
    pgAge: Int = 0,
    genres: List<Genre> = emptyList(),
    runningTime: Int = 0,
    imageUrl: String? = null
): Movie {
    return Movie(
        id = id,
        title = title,
        reviewCount = reviewCount,
        isLiked = isLiked,
        rating = rating,
        pgAge = pgAge,
        genres = genres,
        runningTime = runningTime,
        imageUrl = imageUrl
    )
}
```

## TODO 3.2

- Open `MoviesListItemMapperTest.kt`
- Instead of constructor for `Movie` use `createMovie()` factory method:

```kotlin
val movie = createMovie(
    id = 671039,
    title = "Test 1",
    reviewCount = 200,
    isLiked = true,
    pgAge = 15,
    runningTime = 55,
    imageUrl = "test url"
)
```

## TODO 3.3

> - Instead of `assertEquals` for whole model use one `assertEquals()` for each parameter (it's easier
>   to add new fields in the future):

Author: ```suggestion
- Instead of `assertEquals` for whole model use one `assertEquals()` for each parameter. It lets you add new fields easier in the future.
```

```kotlin
assertEquals(671039, mappedItem.id)
assertEquals("Test 1", mappedItem.title)
assertEquals(200, mappedItem.reviewCount)
assertEquals(true, mappedItem.isLiked)
assertEquals(15, mappedItem.pgAge)
assertEquals(55, mappedItem.runningTime)
assertEquals("test url", mappedItem.imageUrl)
```

## TODO 3.4

- Uncomment `createMapper()` factory method and use it instead of `MoviesListItemMapper` constructor

```kotlin
 private fun createMapper() = MoviesListItemMapper()
```

## TODO 3.5

Because we already tested mapping and created factory method, we can simplify this code for creating
model only with id's.

- Instead of `Movie` constructors use `createMovie(id = [1])` factory methods

```kotlin
val movies = listOf(
    createMovie(id = 1),
    createMovie(id = 2)
)
```

## TODO 3.6

- Remove `mappedMovieList` - now, when we check only `id`'s, whole models are not needed

## TODO 3.7

Now we can check only `id`'s from `viewmodel`'s output and remove previous `assertEquals()` method:

```kotlin
val loadedIds = movieLoadedState.movies.map { it.id }
assertEquals(listOf(1, 2, ...), loadedIds)

> ```

Author: надо б еще объяснить откуда `movieLoadedState` взять. в тесте мы его через assertTrue смарт кастом к нужному типу приводим

        reviewCount = reviewCount,
        isLiked = isLiked,
        rating = rating,
        pgAge = pgAge,
        genres = genres,
        runningTime = runningTime,
        imageUrl = imageUrl
    )
}
```

## TODO 3.2 MoviesListItemMapperTest

- Open `MoviesListItemMapperTest.kt`
- Replace the class name `Movie` with factory method call `createMovie`

Instead of

```kotlin
@Test
fun `mapper maps some fields as is`() {
    ...
    val movie = Movie(...)
    ...
}
```

use

```kotlin
@Test
fun `mapper maps some fields as is`() {
    ...
    val movie = createMovie(...)
    ...
}
```

## TODO 3.3

- Instead of `assertEquals` for whole `MoviesListItem` model use one `assertEquals()` for each
  parameter:

Instead of

```kotlin
@Test
fun `mapper maps some fields as is`() {
    ...
    val expectedMovieListItem = MoviesListItem(
        id = 671039,
        title = "Test 1",
        reviewCount = 200,
        isLiked = true,
        pgAge = 15,
        runningTime = 55,
        imageUrl = "test url",
        genres = listOf(),
        rating = 4
    )
    assertEquals(expectedMovieListItem, listItem)
}
```

use

```kotlin
@Test
fun `mapper maps some fields as is`() {
    ...

>     assertEquals(671039, mappedItem.id)

Reviewer: Название параметра не соотносится с первым воркшопом. Там он называется expectedMovieListItem

Reviewer 2: Поменял на  listItem (expectedMovieListItem больше вообще нет, а вот mappedItem должна называться listItem )

    assertEquals("Test 1", mappedItem.title)
    assertEquals(200, mappedItem.reviewCount)
    assertEquals(true, mappedItem.isLiked)
    assertEquals(15, mappedItem.pgAge)
    assertEquals(55, mappedItem.runningTime)
    assertEquals("test url", mappedItem.imageUrl)
}
```

## TODO 3.4 MoviesListViewModelTest

Because we already tested `MoviesListItemMapper` and created `createMovie(...)` factory method, we
can simplify this code for creating `Movie`'s model only with id.

- Open `MoviesListViewModelTest.kt`
- Instead of `Movie` constructors use `createMovie(id = [1,2,...])` factory methods

Change

```kotlin
@Test
fun `moviesStateOutput by default returns movies list`() {
    val movies = listOf(
        Movie(...),
        Movie(...)
    )
    ...
}
```

to

```kotlin
@Test
fun `moviesStateOutput by default returns movies list`() {
    val movies = listOf(
        createMovie(id = 1),

>         createMovie(id = 2)

Reviewer: Там было 3 movie объекта. Тут два.

Reviewer 2: Из-за рефакторинга  оно не влияет никак, но чисто для консистентности добавил 3й

    )
    ...
}

```

## TODO 3.5

- Remove `val mappedMovieList = listOf(...)` from first test - now, when we check only `id`'s, whole
  list creation is not needed

## TODO 3.6

> Now we can check only `id`'s from viewmodel's output and remove previous `assertEquals()` method:

Reviewer: Тут я не понял что нужно сделать.

Reviewer: Я так понимаю надо удалить         val expectedState = MoviesListViewState.MoviesLoaded(mappedMovieList)
с прошлого воркшопа

Reviewer 2: Добавил сообщение о том что можно удалить его (еще пометил в след шагах, что AssertEquals тоже можно изменить на новый)

- Safely cast viewmodel's output to `MoviesListViewState.MoviesLoaded`

```kotlin
@Test
fun `moviesStateOutput by default returns movies list`() {
    ...

>     val movieLoadedState =

Reviewer: Должны написать что мы заменяем старый val movieLoadedState = viewModel.moviesStateOutput.value на новый     val movieLoadedState =
            viewModel.moviesStateOutput.value as? MoviesListViewState.MoviesLoaded

Reviewer 2: Done

        viewModel.moviesStateOutput.value as? MoviesListViewState.MoviesLoaded
    ...
}

```

- Map `movieLoadedState` to it `id`'s and then use `assertEquals` with simple `listOf(1,2)`

```kotlin
@Test
fun `moviesStateOutput by default returns movies list`() {
    ...
    assertEquals(
        listOf(1, 2),
        movieLoadedState?.movies?.map { it.id }
    )
}
```

## TODO 3.7

- Create factory method for `MoviesListViewModelImpl`

```kotlin
class MoviesListViewModelTest {
    ...

    private fun createMoviesListViewModel(repository: MovieRepository): MoviesListViewModel =
        MoviesListViewModelImpl(repository, MoviesListItemMapper())
}
```

- Use factory instead of `MoviesListViewModelImpl` constructor:

Change

```kotlin
@Test
fun `moviesStateOutput by default returns movies list`() {
    ...
    val viewModel = MoviesListViewModelImpl(repository, MoviesListItemMapper())
    ...
}
```

to

```kotlin
@Test
fun `moviesStateOutput by default returns movies list`() {
    ...

>     val viewModel = createMoviesListViewModel(repository)

Reviewer: Это в двух тестах надо поменять как я понимаю.

Reviewer 2: Я добавил сообщение, оно почти такое же(только название метода различается), но нужно будет субъективно глянуть, не замного ли текста

# Workshop 3

- Now we can refactor and simplify code from Workshop 1-2

## TODO 3.1

> - Create `MovieFactory.kt` file in `test/****/utils` package

Reviewer 3: Тут нужно, или прописать полное название пакета и попросить студентов создать его, или просто сказать "разместите рядом с тестами в директории `test`"

Author: уточнение. мы factory methods помещали в тот же пакет где и модель в основном коде. 
`package com.android.academy.fundamentals.homework.model`

Reviewer 3: Я к тому, что нужно указать полное название пакета вместо `com.android.academy.fundamentals.homework.model`

Reviewer 2: Тут я пока ничего не менял, потому что до конца не понял, какую именно папку указывать. Обсудим

Reviewer 2: Summary: расписать более конкретно:
1. Создайте пакет, создайте файл, создайте метод

- Create or copy-paste `createMovie(**)` factory method

```kotlin
fun createMovie(
    id: Int = 0,
    title: String = "",
    reviewCount: Int = 0,
    isLiked: Boolean = false,
    rating: Int = 0,
    pgAge: Int = 0,
    genres: List<Genre> = emptyList(),
    runningTime: Int = 0,
    imageUrl: String? = null
): Movie {
    return Movie(
        id = id,
        title = title,
        reviewCount = reviewCount,
        isLiked = isLiked,
        rating = rating,
        pgAge = pgAge,
        genres = genres,
        runningTime = runningTime,
        imageUrl = imageUrl
    )
}
```

## TODO 3.2 MoviesListItemMapperTest

- Open `MoviesListItemMapperTest.kt`

> - Instead of constructor for `Movie` use `createMovie()` factory method:

Reviewer 3: Replace the class name `Movie` with factory method call `createMovie`

Reviewer 2: Done

  Instead of

```kotlin
val movie = Movie(**)
```

use

```kotlin
val movie = createMovie(
    id = 671039,
    title = "Test 1",
    reviewCount = 200,
    isLiked = true,
    pgAge = 15,
    runningTime = 55,
    imageUrl = "test url"
)
```

## TODO 3.3

- Instead of `assertEquals` for whole `MoviesListItem` model use one `assertEquals()` for each
  parameter (it's easier to add new fields in the future):

> - Instead of

Reviewer 3: На мой взгляд, маркер "-" тут лишний.

Reviewer 2: Done

```kotlin
val expectedMovieListItem = MoviesListItem(
    id = 671039,
    title = "Test 1",
    reviewCount = 200,
    isLiked = true,
    pgAge = 15,
    runningTime = 55,
    imageUrl = "test url",
    genres = listOf(),
    rating = 4
)
```

use

```kotlin
assertEquals(671039, mappedItem.id)
assertEquals("Test 1", mappedItem.title)
assertEquals(200, mappedItem.reviewCount)
assertEquals(true, mappedItem.isLiked)
assertEquals(15, mappedItem.pgAge)
assertEquals(55, mappedItem.runningTime)
assertEquals("test url", mappedItem.imageUrl)
```

> It lets you add new fields easier in the future.

Reviewer 3: Уже упоминалось в строке 63.

Reviewer 2: Оу, я умудрился весь воркшоп продублировать  - уже убрал, оставил только один

## TODO 3.4 MoviesListViewModelTest

- Open `MoviesListViewModelTest.kt`
- Create `createMapper()` factory method inside of `MoviesListViewModelTest.kt` and use it instead
  of `MoviesListItemMapper` constructor

```kotlin
 private fun createMapper() = MoviesListItemMapper()

> ```

Reviewer 3: Нужен краткий пример (big picture).

Reviewer 2: Я пересмотрел воркшоп - и там в целом для маппера не нужен фактори метод, потому что потом используется фактори метод для всей вьюмодели
Я убрал этот шаг полностью

- Instead of constructor for `Movie` use `createMovie()` factory method:
  Instead of

```kotlin
val movie = Movie(**)
```

use

```kotlin
val movie = createMovie(
    id = 671039,
    title = "Test 1",
    reviewCount = 200,
    isLiked = true,
    pgAge = 15,
    runningTime = 55,
    imageUrl = "test url"
)
```

## TODO 3.3

- Instead of `assertEquals` for whole `MoviesListItem` model use one `assertEquals()` for each
  parameter (it's easier to add new fields in the future):

- Instead of

```kotlin
val expectedMovieListItem = MoviesListItem(
    id = 671039,
    title = "Test 1",
    reviewCount = 200,
    isLiked = true,
    pgAge = 15,
    runningTime = 55,
    imageUrl = "test url",
    genres = listOf(),
    rating = 4
)
```

use

```kotlin
assertEquals(671039, mappedItem.id)
assertEquals("Test 1", mappedItem.title)
assertEquals(200, mappedItem.reviewCount)
assertEquals(true, mappedItem.isLiked)
assertEquals(15, mappedItem.pgAge)
assertEquals(55, mappedItem.runningTime)
assertEquals("test url", mappedItem.imageUrl)
```

It lets you add new fields easier in the future.

## TODO 3.4 MoviesListViewModelTest

- Open `MoviesListViewModelTest.kt`

> - Create `createMapper()` factory method inside of `MoviesListViewModelTest.kt` and use it instead

Reviewer 3: По-моему, это - что-то лишнее.

Reviewer 2: Да, убрал полностью этот шаг в итоге

```

## TODO 3.5

Because we already tested `MoviesListItemMapper` and created `createMovie(***)` factory method, we
can simplify this code for creating `Movie`'s model only with id.

- Instead of `Movie` constructors use `createMovie(id = [1,2,...])` factory methods

```kotlin
val movies = listOf(
    createMovie(id = 1),
    createMovie(id = 2)
)
```

## TODO 3.6

- Remove `val mappedMovieList = listOf(***)` - now, when we check only `id`'s, whole list creation
  is not needed

## TODO 3.7

Now we can check only `id`'s from viewmodel's output and remove previous `assertEquals()` method:

- Safely cast viewmodel's output to `MoviesListViewState.MoviesLoaded`

```kotlin
val movieLoadedState = viewModel.moviesStateOutput.value as? MoviesListViewState.MoviesLoaded
```

- Map `movieLoadedState` to it `id`'s and then use `assertEquals` with simple `listOf(1,2)`

```kotlin
assertEquals(
    listOf(1, 2),
    movieLoadedState?.movies?.map { it.id }
)
```

## TODO 3.8

- Create factory method for `viewModel`

```kotlin
  private fun createMoviesListViewModel(repository: MovieRepository): MoviesListViewModel =
    MoviesListViewModelImpl(repository, MoviesListItemMapper())
```

> - Use factory instead of `MoviesListViewModelImpl` constructor:

Reviewer 3: Нужно бы добавить пример.

Reviewer 2: done

## TODO 3.9

- Run all refactored tests to check them works

# Summary

> - `MoviesListItemMapperTest.kt`

Reviewer 3: Кажется, стоит заменить на заголовок второго уровня

Reviewer 2: done

```kotlin
package com.android.academy.fundamentals.homework.presentation.features.movies.viewmodel

import com.android.academy.fundamentals.homework.model.createMovie
import org.junit.Assert.assertEquals
import org.junit.Test

class MoviesListItemMapperTest {

    @Test
    fun `mapper maps some fields as is`() {
        val mapper = MoviesListItemMapper()
        val movie = createMovie(
            id = 671039,
            title = "Test 1",
            reviewCount = 200,
            isLiked = true,
            pgAge = 15,
            runningTime = 55,
            imageUrl = "test url",
            genres = listOf(),
            rating = 4
        )

        val listItem = mapper.map(movie)

        assertEquals(671039, listItem.id)
        assertEquals("Test 1", listItem.title)
        assertEquals(200, listItem.reviewCount)
        assertEquals(true, listItem.isLiked)
        assertEquals(15, listItem.pgAge)
        assertEquals(55, listItem.runningTime)
        assertEquals("test url", listItem.imageUrl)
    }
}
```

> - `MoviesListViewModelTest.kt`

Reviewer 3: Кажется, стоит заменить на заголовок второго уровня

Reviewer 2: done
